step/s3.go

```diff
@@ -0,0 +1,378 @@
+package step
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/smithy-go"
+	"github.com/bitrise-io/go-steputils/v2/cache/network"
+	"github.com/bitrise-io/go-utils/retry"
+	"github.com/bitrise-io/go-utils/v2/log"
+)
+
+const (
+	numUploadRetries = 3
+	maxKeyLength     = 512
+	maxKeyCount      = 8
+
+	// 50MB
+	multipartChunkSize = 50_000_000
+
+	// 100MB
+	copySizeLimit = 100_000_000
+
+	// The archive checksum is uploaded with the object as metadata.
+	// As we can be pretty sure of the integrity of the uploaded archive,
+	// so we avoid having to locally pre-calculate the SHA-256 checksum of each (multi)part.
+	// Instead we include the full object checksum as metadata, and compare against it at consecutive uploads.
+	checksumKey = "full-object-checksum-sha256"
+)
+
+// UploadService .
+type UploadService struct {
+	Client          *s3.Client
+	Bucket          string
+	Region          string
+	AccessKeyID     string
+	SecretAccessKey string
+}
+
+// Upload .
```

Comment:
nit: can be removed, steps are linted with different rules than go-steplib, comments are not required for exported functions.